compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts

```diff
@@ -0,0 +1,512 @@
+  #declarations: Map<DeclarationId, Decl> = new Map();
+  #reassignments: Map<Identifier, Decl> = new Map();
+  // Reactive dependencies used in the current reactive scope.
+  #dependencies: Stack<Array<ReactiveScopeDependency>> = empty();
+  /*
+   * We keep a sidemap for temporaries created by PropertyLoads, and do
+   * not store any control flow (i.e. #inConditionalWithinScope) here.
+   *  - a ReactiveScope (A) containing a PropertyLoad may differ from the
+   *    ReactiveScope (B) that uses the produced temporary.
+   *  - codegen will inline these PropertyLoads back into scope (B)
+   */
+  #properties: ReadonlyMap<Identifier, ReactiveScopeDependency>;
+  #temporaries: ReadonlyMap<Identifier, Identifier>;
+  #scopes: Stack<ReactiveScope> = empty();
+  deps: Map<ReactiveScope, Array<ReactiveScopeDependency>> = new Map();
+
+  constructor(
+    temporariesUsedOutsideScope: Set<DeclarationId>,
+    temporaries: ReadonlyMap<Identifier, Identifier>,
+    properties: ReadonlyMap<Identifier, ReactiveScopeDependency>,
+  ) {
+    this.#temporariesUsedOutsideScope = temporariesUsedOutsideScope;
+    this.#temporaries = temporaries;
+    this.#properties = properties;
+  }
+
+  enterScope(scope: ReactiveScope): void {
+    // Set context for new scope
+    this.#dependencies = this.#dependencies.push([]);
+    this.#scopes = this.#scopes.push(scope);
+  }
+
+  exitScope(scope: ReactiveScope, pruned: boolean): void {
+    // Save dependencies we collected from the exiting scope
+    const scopedDependencies = this.#dependencies.value;
+    CompilerError.invariant(scopedDependencies != null, {
+      reason: '[PropagateScopeDeps]: Unexpected scope mismatch',
+      loc: scope.loc,
+    });
+
+    // Restore context of previous scope
+    this.#scopes = this.#scopes.pop();
+    this.#dependencies = this.#dependencies.pop();
+
+    /*
+     * Collect dependencies we recorded for the exiting scope and propagate
+     * them upward using the same rules as normal dependency collection.
+     * Child scopes may have dependencies on values created within the outer
+     * scope, which necessarily cannot be dependencies of the outer scope.
+     */
+    for (const dep of scopedDependencies) {
+      if (this.#checkValidDependency(dep)) {
+        this.#dependencies.value?.push(dep);
+      }
+    }
+
+    if (!pruned) {
+      this.deps.set(scope, scopedDependencies);
+    }
+  }
+
+  isUsedOutsideDeclaringScope(place: Place): boolean {
+    return this.#temporariesUsedOutsideScope.has(
+      place.identifier.declarationId,
+    );
+  }
+
+  /*
+   * Records where a value was declared, and optionally, the scope where the value originated from.
+   * This is later used to determine if a dependency should be added to a scope; if the current
+   * scope we are visiting is the same scope where the value originates, it can't be a dependency
+   * on itself.
+   */
+  declare(identifier: Identifier, decl: Decl): void {
```

Comment:
All below methods are copied from propagateScopeDeps (reactive scopes version). The only change is that I cleaned up all `isConditional` / non-ast control flow (return / throw) logic, and moved temporaries / properties sidemap creation to an earlier step